Content.Shared/_Crescent/Hardpoints/GridHardpointLimiterSystem.cs

```diff
@@ -0,0 +1,221 @@
+        SubscribeLocalEvent<MapGridComponent, MapInitEvent>(OnGridMapInit);
+
+        // IMPORTANT: Listen for hardpoint anchoring attempts
+        SubscribeLocalEvent<HardpointComponent, AnchorAttemptEvent>(OnHardpointAnchorAttempt);
+        SubscribeLocalEvent<HardpointComponent, UnanchorAttemptEvent>(OnHardpointUnanchorAttempt);
+
+        // CRITICAL: Final check right before anchoring completes
+        SubscribeLocalEvent<HardpointComponent, BeforeAnchoredEvent>(OnHardpointBeforeAnchored);
+    }
+
+    private void OnGridMapInit(EntityUid uid, MapGridComponent component, MapInitEvent args)
+    {
+        // Add a small delay to ensure all entities are loaded
+        Timer.Spawn(200, () => InitializeGridLimits(uid));
+    }
+
+    private void OnHardpointAnchorAttempt(EntityUid uid, HardpointComponent component, AnchorAttemptEvent args)
+    {
+ //       Logger.Info($"[HardpointLimiter] OnHardpointAnchorAttempt called for hardpoint {uid}");
+
+        var gridUid = Transform(uid).GridUid;
+        if (gridUid == null)
+        {
+//            Logger.Warning($"[HardpointLimiter] Hardpoint {uid} has no grid, allowing anchor");
+            return;
+        }
+
+ //       Logger.Info($"[HardpointLimiter] Checking if hardpoint {uid} can be anchored to grid {gridUid}");
+
+        if (!CanInstall(gridUid.Value))
+        {
+//            Logger.Warning($"[HardpointLimiter] BLOCKING hardpoint {uid} anchor - limit exceeded on grid {gridUid}");
+            args.Cancel();
+            return;
+        }
+
+ //       Logger.Info($"[HardpointLimiter] Allowing hardpoint {uid} to anchor to grid {gridUid}");
+    }
+
+    private void OnHardpointBeforeAnchored(EntityUid uid, HardpointComponent component, BeforeAnchoredEvent args)
+    {
+//        Logger.Info($"[HardpointLimiter] OnHardpointBeforeAnchored called for hardpoint {uid} (FINAL CHECK)");
+
+        var gridUid = Transform(uid).GridUid;
+        if (gridUid == null)
+        {
+//            Logger.Warning($"[HardpointLimiter] Hardpoint {uid} has no grid during final check, allowing anchor");
+            return;
+        }
+
+//        Logger.Info($"[HardpointLimiter] FINAL CHECK: Can hardpoint {uid} be anchored to grid {gridUid}?");
+
+        if (!CanInstall(gridUid.Value))
+        {
+//            Logger.Warning($"[HardpointLimiter] FINAL CHECK FAILED: BLOCKING hardpoint {uid} anchor - limit exceeded on grid {gridUid}");
+
+            // We can't cancel BeforeAnchoredEvent, so we need to unanchor it immediately
+            // This is a bit hacky but necessary since BeforeAnchoredEvent isn't cancellable
+            Timer.Spawn(1, () => {
+                if (Exists(uid) && Transform(uid).Anchored)
+                {
+ //                   Logger.Warning($"[HardpointLimiter] Force-unanchoring hardpoint {uid} due to limit exceeded");
+                    _xform.Unanchor(uid);
+                }
+            });
```

Comment:
This has to go , hook ono the event that the hardpoint component uses for allowing a hardpointGun anchor instead so you can actually cancel it